File: src/other/_208_trie.rs
#[derive(Default)]
struct Trie {
    child: [Option<Box<Trie>>; 26],
    is_end: bool,
}

impl Trie {
    fn new() -> Self {
        Trie::default()
    }

    fn insert(&mut self, word: String) {
        let mut cur = self;
        for i in word.chars().map(|ch| (ch as u8 - 'a' as u8) as usize) {
            cur = cur.child[i].get_or_insert_with(|| Box::new(Trie::new()));
        }
        cur.is_end = true;
    }

    fn search(&self, word: String) -> bool {
        let mut cur = self;
        for i in word.chars().map(|ch| (ch as u8 - 'a' as u8) as usize) {
            match cur.child[i].as_ref() {
                Some(node) => cur = node,
                None => return false,
            }
        }
        cur.is_end
    }

    fn starts_with(&self, prefix: String) -> bool {
        let mut cur = self;
        for i in prefix.chars().map(|ch| (ch as u8 - 'a' as u8) as usize) {
            match cur.child[i].as_ref() {
                Some(node) => cur = node,
                None => return false,
            }
        }
        true
    }
}

File: hot-100/208_trie.go
package hot_100

type Trie struct {
	child [26]*Trie
	end   bool
}

func Constructor2() Trie {
	return Trie{}
}

func (this *Trie) Insert(word string) {
	node := this
	for _, ch := range word {
		ch -= 'a'
		if node.child[ch] == nil {
			node.child[ch] = &Trie{}
		}
		node = node.child[ch]
	}
	node.end = true
}

func (this *Trie) SearchPrefix(prefix string) *Trie {
	node := this
	for _, ch := range prefix {
		ch -= 'a'
		if node.child[ch] == nil {
			return nil
		}
		node = node.child[ch]
	}
	return node
}

func (this *Trie) Search(word string) bool {
	node := this.SearchPrefix(word)
	return node != nil && node.end
}

func (this *Trie) StartsWith(prefix string) bool {
	return this.SearchPrefix(prefix) != nil
}
